motionplan/planManager.go

```diff
@@ -0,0 +1,382 @@
+					remainingSteps, err = mp.planMotion(ctx, goals, nextSeed.Q(), opts, nil, iter+1)
+					if err != nil {
+						return nil, err
+					}
+				}
+				for {
+					// Get the step from this runner invocation, and return everything in order.
+					select {
+					case <-ctx.Done():
+						return nil, ctx.Err()
+					default:
+					}
+
+					select {
+					case finalSteps := <-solutionChan:
+						if finalSteps.err() != nil {
+							return nil, finalSteps.err()
+						}
+						results := append(finalSteps.toInputs(), remainingSteps...)
+						return results, nil
+					default:
+					}
+				}
+			case finalSteps := <-solutionChan:
+				// We didn't get a solution preview (possible error), so we get and process the full step set and error.
+				if finalSteps.err() != nil {
+					if opt.Fallback != nil {
+						remainingSteps, err = mp.planMotion(
+							ctx,
+							goals,
+							seed,
+							append([]*plannerOptions{opt.Fallback}, opts[1:]...),
+							finalSteps.rm,
+							iter,
+						)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						return nil, finalSteps.err()
+					}
+				} else {
+					if iter+1 < len(goals) {
+						// in this case, we create the next step (and thus the remaining steps) and the
+						// step from our iteration hangs out in the channel buffer until we're done with it
+						remainingSteps, err = mp.planMotion(
+							ctx,
+							goals,
+							finalSteps.steps[len(finalSteps.steps)-1].Q(),
+							opts,
+							nil,
+							iter+1,
+						)
+						if err != nil {
+							return nil, err
+						}
+					}
+				}
+				results := append(finalSteps.toInputs(), remainingSteps...)
+				return results, nil
+			default:
+			}
+		}
+	} else {
+		resultSlicesRaw, err := pathPlanner.Plan(planctx, goal, seed, opt)
+		if err != nil {
+			return nil, err
+		}
+		if iter < len(goals)-2 {
+			// in this case, we create the next step (and thus the remaining steps)
+			remainingSteps, err = mp.planMotion(ctx, goals, resultSlicesRaw[len(resultSlicesRaw)-1], opts, nil, iter+1)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return append(resultSlicesRaw, remainingSteps...), nil
+	}
+}
+
+// This is where the map[string]interface{} passed in via `extra` is used to decide how planning happens.
+func (mp *planManager) plannerSetupFromMoveRequest(
+	from, to spatialmath.Pose,
+	seedMap map[string][]referenceframe.Input,
+	worldState *commonpb.WorldState,
+	planningOpts map[string]interface{},
+) (*plannerOptions, error) {
+	// Start with normal options
+	opt := newBasicPlannerOptions()
+
+	opt.extra = planningOpts
+
+	collisionConstraint, err := NewCollisionConstraintFromWorldState(mp.frame, mp.fs, worldState, seedMap)
+	if err != nil {
+		return nil, err
+	}
+	opt.AddConstraint(defaultCollisionConstraintName, collisionConstraint)
+
+	// error handling around extracting motion_profile information from map[string]interface{}
+	var motionProfile string
+	profile, ok := planningOpts["motion_profile"]
```

Comment:
Could this be consolidated down to a single `if motionProfile, ok := planningOpts["motion_profile"].(string); ok {` line?

Reply:
Not quite; if "motion_profile" is not set, we want to continue. This is just checking that if it is set, it's parseable.

Reply:
OK, very good.